File: pipeline/stages/score.py
"""Score stage — aggregate extractions into per-restaurant scores.

For each restaurant in a city:
  - food_score = 0.75 * positive_rate + 0.25 * pos_neg_ratio_normalized
  - service_score = same shape
  - food_unique_users = distinct comments that supplied a food sentiment
  - service_unique_users = distinct comments that supplied a service sentiment
  - total_unique_users = distinct comments touching either aspect

Then sort restaurants by a Bayesian-smoothed food score (penalizes "100% liked
it" with N=1 against e.g. "92% liked it" with N=80) and write city_rank.

Idempotent: upserts into `restaurant_scores` keyed on restaurant_id.

Note on "unique users": we use comment_id as the proxy because the same
Reddit user posting multiple comments would still produce multiple
extractions tied to different comments. Tying back to comment.author for
true per-author dedup is a future refinement.
"""
from __future__ import annotations

from typing import Optional

from pipeline import db


# ---------- public API ------------------------------------------------------


def compute_scores_for_city(city_slug: str) -> int:
    """Recompute and persist scores for every restaurant in a city.

    Returns the number of restaurants scored.
    """
    client = db.get_client()
    restaurants = (
        client.table("restaurants")
        .select("id")
        .eq("city_slug", city_slug)
        .eq("closed", False)
        .execute()
    )

    scored: list[dict] = []
    for r in restaurants.data or []:
        agg = _aggregate_for_restaurant(r["id"])
        if agg["total_unique_users"] == 0:
            # No extractions touch this restaurant — skip; we'd otherwise
            # write a row of zeros that ranks dead-last and clutters the UI.
            continue
        scored.append({"restaurant_id": r["id"], **agg})

    # Restaurants whose ONLY signal is negative (no positive votes on either
    # aspect) get city_rank=NULL so they fall out of the public list. The
    # row stays so /admin can still inspect them, and they'll auto-reappear
    # once a positive mention surfaces in a later pipeline run.
    rankable = [s for s in scored if not _is_only_negative(s)]
    rankable.sort(key=lambda s: (-_rank_score(s), -s["total_unique_users"]))
    for i, s in enumerate(rankable, start=1):
        s["city_rank"] = i
    for s in scored:
        if _is_only_negative(s):
            s["city_rank"] = None

    if scored:
        client.table("restaurant_scores").upsert(
            scored, on_conflict="restaurant_id"
        ).execute()

    return len(rankable)


# ---------- internals -------------------------------------------------------


TAG_STICK_THRESHOLD = 2  # Min extractions that must reference a tag for it to stick.


def _aggregate_for_restaurant(restaurant_id: str) -> dict:
    client = db.get_client()
    rows = (
        client.table("extractions")
        .select("food_sentiment, service_sentiment, vote_weight, comment_id, tags")
        .eq("restaurant_id", restaurant_id)
        .execute()
        .data
        or []
    )

    food_pos = food_neg = 0.0
    service_pos = service_neg = 0.0
    food_users: set[str] = set()
    service_users: set[str] = set()
    tag_counts: dict[str, int] = {}

    for row in rows:
        w = float(row.get("vote_weight") or 1.0)
        comment_id = row["comment_id"]

        food = row.get("food_sentiment")
        if food is not None:
            food_users.add(comment_id)
            if food == "positive":
                food_pos += w
            elif food == "negative":
                food_neg += w
            elif food == "mixed":
                food_pos += w * 0.5
                food_neg += w * 0.5

        service = row.get("service_sentiment")
        if service is not None:
            service_users.add(comment_id)
            if service == "positive":
                service_pos += w
            elif service == "negative":
                service_neg += w
            elif service == "mixed":
                service_pos += w * 0.5
                service_neg += w * 0.5

        for tag in (row.get("tags") or []):
            tag_counts[tag] = tag_counts.get(tag, 0) + 1

    sticky_tags = sorted(
        t for t, n in tag_counts.items() if n >= TAG_STICK_THRESHOLD
    )

    return {
        "food_score":           _aspect_score(food_pos, food_neg),
        "food_positive":        round(food_pos, 3),
        "food_negative":        round(food_neg, 3),
        "food_unique_users":    len(food_users),
        "service_score":        _aspect_score(service_pos, service_neg),
        "service_positive":     round(service_pos, 3),
        "service_negative":     round(service_neg, 3),
        "service_unique_users": len(service_users),
        "total_unique_users":   len(food_users | service_users),
        "tags":                 sticky_tags,
    }


SCORE_PRIOR_ALPHA = 2.0   # Pseudo-positive votes baked in.
SCORE_PRIOR_BETA = 1.5    # Pseudo-negative votes baked in.
# Neutral prior = α / (α + β) ≈ 0.571 (~5.7 on the displayed 0–10 scale).
# The slight positive lean reflects that most Reddit mentions of a
# restaurant are recommendations rather than pans, so a single mention
# with no other context should land slightly above the midline.


def _aspect_score(positive: float, negative: float) -> Optional[float]:
    """Beta(α=2, β=1.5)-smoothed positive rate, returned in [0, 1].

    Returns None when there's no signal at all (caller renders "no data").
    Otherwise:
      - 1 positive, 0 negative → 0.667 (~6.7 on the 0–10 scale)
      - 50 positive, 0 negative → 0.972 (~9.7)
      - 1 negative, 0 positive → 0.444 (~4.4)
      - 50 negative, 0 positive → 0.037 (~0.4)
      - 5 positive, 5 negative → 0.519 (~5.2)
    Reaching 1.0 or 0.0 is effectively impossible at any sample size,
    which keeps the displayed score honest about confidence.
    """
    if positive + negative == 0:
        return None
    return round(
        (positive + SCORE_PRIOR_ALPHA)
        / (positive + negative + SCORE_PRIOR_ALPHA + SCORE_PRIOR_BETA),
        3,
    )


def _is_only_negative(s: dict) -> bool:
    """True when a restaurant has negative votes but zero positive votes on
    BOTH aspects — i.e. every Reddit mention was a pan. We keep these rows
    in restaurant_scores (admin needs visibility) but null out city_rank so
    the public list filter excludes them.

    Mixed sentiments contribute 0.5 to positive, so a "mixed" review keeps
    a restaurant rankable.
    """
    food_pos = float(s.get("food_positive") or 0)
    service_pos = float(s.get("service_positive") or 0)
    food_neg = float(s.get("food_negative") or 0)
    service_neg = float(s.get("service_negative") or 0)
    return (food_pos + service_pos) == 0 and (food_neg + service_neg) > 0


def _rank_score(s: dict) -> float:
    """Sort key for the ranked list. The Beta prior in `_aspect_score` already
    pulls low-N scores toward the midline, so the rank score is just the
    food score itself. Total-volume tie-breaks live in the sort tuple.
    """
    return float(s.get("food_score") or 0.0)

File: lib/score.ts
import "server-only";
import { adminClient } from "./supabase-admin";

/**
 * TypeScript port of pipeline/stages/score.py::compute_scores_for_city.
 *
 * Used by the /admin "Recompute scores" button so admins can refresh
 * /[city] immediately after dismissing/reassigning flags, without waiting
 * for the next Python pipeline run. Same algorithm as the pipeline:
 *
 *   food_score    = 0.75 * positive_rate + 0.25 * pos/neg_ratio_normalized
 *   service_score = same shape
 *   rank_score    = food_score * (food_n / (food_n + 5))   (Bayesian smoothing)
 *
 * If you change the formula here, change it in pipeline/stages/score.py too.
 */

const TAG_STICK_THRESHOLD = 2;

type ExtractionRow = {
  restaurant_id: string | null;
  food_sentiment: "positive" | "negative" | "mixed" | null;
  service_sentiment: "positive" | "negative" | "mixed" | null;
  vote_weight: number | string;
  comment_id: string;
  tags: string[] | null;
};

type Aggregate = {
  food_score: number | null;
  food_positive: number;
  food_negative: number;
  food_unique_users: number;
  service_score: number | null;
  service_positive: number;
  service_negative: number;
  service_unique_users: number;
  /** Distinct comments that named the place but supplied NEITHER sentiment
   *  (bare-name responses in neutral search threads — see extract.py rule 6).
   *  Volume signal only; doesn't affect food_score or service_score. */
  mention_only_users: number;
  total_unique_users: number;
  tags: string[];
};

export async function computeScoresForCity(citySlug: string): Promise<number> {
  const supabase = adminClient();

  // 1. Fetch open restaurants for this city
  const { data: restaurants, error: rErr } = await supabase
    .from("restaurants")
    .select("id")
    .eq("city_slug", citySlug)
    .eq("closed", false);
  if (rErr) throw new Error(`Failed to fetch restaurants: ${rErr.message}`);
  if (!restaurants || restaurants.length === 0) return 0;

  const restaurantIds = restaurants.map((r) => r.id as string);

  // 2. Fetch all extractions linked to these restaurants. Chunk by 100 to
  //    keep the IN-clause URL under PostgREST's request-line limit.
  const extractions: ExtractionRow[] = [];
  const CHUNK = 100;
  for (let i = 0; i < restaurantIds.length; i += CHUNK) {
    const chunk = restaurantIds.slice(i, i + CHUNK);
    const { data, error } = await supabase
      .from("extractions")
      .select("restaurant_id, food_sentiment, service_sentiment, vote_weight, comment_id, tags")
      .in("restaurant_id", chunk);
    if (error) throw new Error(`Failed to fetch extractions: ${error.message}`);
    if (data) extractions.push(...(data as unknown as ExtractionRow[]));
  }

  // 3. Aggregate per restaurant. We keep the unique-user Sets in a parallel
  //    map so they never leak into the upsert payload — PostgREST would
  //    reject unknown columns ("Could not find the '_foodUsers' column…").
  const byRestaurant = new Map<string, Aggregate>();
  const usersByRestaurant = new Map<string, UserTracking>();
  for (const row of extractions) {
    if (!row.restaurant_id) continue;
    let agg = byRestaurant.get(row.restaurant_id);
    let users = usersByRestaurant.get(row.restaurant_id);
    if (!agg) {
      agg = newAggregate();
      byRestaurant.set(row.restaurant_id, agg);
    }
    if (!users) {
      users = {
        food: new Set<string>(),
        service: new Set<string>(),
        mentionOnly: new Set<string>(),
        tagCounts: new Map<string, number>(),
      };
      usersByRestaurant.set(row.restaurant_id, users);
    }
    accumulate(agg, users, row);
  }

  // Finalize: compute scores from the running totals, drop zero-volume rows.
  const scored: Array<
    { restaurant_id: string } & Aggregate & { city_rank: number | null }
  > = [];
  for (const [id, agg] of byRestaurant) {
    if (agg.total_unique_users === 0) continue;
    agg.food_score = aspectScore(agg.food_positive, agg.food_negative);
    agg.service_score = aspectScore(agg.service_positive, agg.service_negative);
    const counts = usersByRestaurant.get(id)?.tagCounts ?? new Map();
    agg.tags = [...counts.entries()]
      .filter(([, n]) => n >= TAG_STICK_THRESHOLD)
      .map(([t]) => t)
      .sort();
    scored.push({ restaurant_id: id, ...agg, city_rank: null });
  }

  // 4. Rank by Bayesian-smoothed food score, then total volume. Restaurants
  //    whose ONLY signal is negative (zero positive on both aspects) get
  //    city_rank=null so they fall out of the public list — but we still
  //    upsert their score row so /admin can see the data.
  const rankable = scored.filter((s) => !isOnlyNegative(s));
  rankable.sort((a, b) => {
    const rA = rankScore(a);
    const rB = rankScore(b);
    if (rA !== rB) return rB - rA;
    return b.total_unique_users - a.total_unique_users;
  });
  rankable.forEach((s, i) => {
    s.city_rank = i + 1;
  });

  // 5. Upsert into restaurant_scores. Chunk to keep the request body sane.
  if (scored.length > 0) {
    const BATCH = 100;
    for (let i = 0; i < scored.length; i += BATCH) {
      const batch = scored.slice(i, i + BATCH);
      const { error } = await supabase
        .from("restaurant_scores")
        .upsert(batch, { onConflict: "restaurant_id" });
      if (error) throw new Error(`Failed to upsert scores: ${error.message}`);
    }
  }

  return rankable.length;
}

// ---------- internals -------------------------------------------------------

type UserTracking = {
  food: Set<string>;
  service: Set<string>;
  mentionOnly: Set<string>;
  tagCounts: Map<string, number>;
};

function newAggregate(): Aggregate {
  return {
    food_score: null,
    food_positive: 0,
    food_negative: 0,
    food_unique_users: 0,
    service_score: null,
    service_positive: 0,
    service_negative: 0,
    service_unique_users: 0,
    mention_only_users: 0,
    total_unique_users: 0,
    tags: [],
  };
}

function accumulate(agg: Aggregate, users: UserTracking, row: ExtractionRow) {
  const w = Number(row.vote_weight) || 1.0;

  if (row.food_sentiment !== null) {
    users.food.add(row.comment_id);
    if (row.food_sentiment === "positive") agg.food_positive += w;
    else if (row.food_sentiment === "negative") agg.food_negative += w;
    else if (row.food_sentiment === "mixed") {
      agg.food_positive += w * 0.5;
      agg.food_negative += w * 0.5;
    }
  }
  if (row.service_sentiment !== null) {
    users.service.add(row.comment_id);
    if (row.service_sentiment === "positive") agg.service_positive += w;
    else if (row.service_sentiment === "negative") agg.service_negative += w;
    else if (row.service_sentiment === "mixed") {
      agg.service_positive += w * 0.5;
      agg.service_negative += w * 0.5;
    }
  }

  // Volume-only: comment named the place but expressed no sentiment.
  // Counted in total_unique_users for rank tiebreak; surfaced on the
  // restaurant card as "+ N more mentions".
  if (row.food_sentiment === null && row.service_sentiment === null) {
    users.mentionOnly.add(row.comment_id);
  }

  for (const tag of row.tags ?? []) {
    users.tagCounts.set(tag, (users.tagCounts.get(tag) ?? 0) + 1);
  }

  // Derived counts in lock-step.
  agg.food_unique_users = users.food.size;
  agg.service_unique_users = users.service.size;
  agg.mention_only_users = users.mentionOnly.size;
  agg.total_unique_users = new Set([
    ...users.food,
    ...users.service,
    ...users.mentionOnly,
  ]).size;

  // Canonical 3-decimal precision (matches Python).
  agg.food_positive = round3(agg.food_positive);
  agg.food_negative = round3(agg.food_negative);
  agg.service_positive = round3(agg.service_positive);
  agg.service_negative = round3(agg.service_negative);
}

function isOnlyNegative(s: Aggregate): boolean {
  // Negative votes exist on at least one aspect AND zero positive votes on
  // both aspects. Mixed sentiments split 0.5 to positive, so a "mixed"
  // review keeps the restaurant rankable.
  const pos = s.food_positive + s.service_positive;
  const neg = s.food_negative + s.service_negative;
  return pos === 0 && neg > 0;
}

// Beta(α=2, β=1.5) prior — slight positive lean since Reddit mentions of a
// restaurant skew toward recommendations. Neutral prior ≈ 0.571 (~5.7 / 10).
// Mirrors pipeline/stages/score.py — keep them in sync.
const SCORE_PRIOR_ALPHA = 2.0;
const SCORE_PRIOR_BETA = 1.5;

function aspectScore(positive: number, negative: number): number | null {
  if (positive + negative === 0) return null;
  return round3(
    (positive + SCORE_PRIOR_ALPHA) /
      (positive + negative + SCORE_PRIOR_ALPHA + SCORE_PRIOR_BETA),
  );
}

function rankScore(s: Aggregate): number {
  // The Beta prior in aspectScore already shrinks low-N scores toward the
  // midline, so ranking just sorts by food_score. Volume tie-break lives
  // in the comparator.
  return s.food_score ?? 0;
}

function round3(n: number): number {
  return Math.round(n * 1000) / 1000;
}
